jolt-core/src/r1cs/builder.rs

```diff
@@ -763,38 +773,55 @@ impl<F: JoltField, I: ConstraintInput> CombinedUniformBuilder<F, I> {
-        let cz_chunks = Cz.par_chunks_mut(self.uniform_repeat);
-
-        self.uniform_builder
-            .constraints
-            .par_iter()
-            .zip(az_chunks.zip(bz_chunks.zip(cz_chunks)))
-            .for_each(|(constraint, (az_chunk, (bz_chunk, cz_chunk)))| {
-                let a_inputs = batch_inputs(&constraint.a);
-                let b_inputs = batch_inputs(&constraint.b);
-                let c_inputs = batch_inputs(&constraint.c);
-
-                constraint.a.evaluate_batch_mut(&a_inputs, az_chunk);
-                constraint.b.evaluate_batch_mut(&b_inputs, bz_chunk);
-                constraint.c.evaluate_batch_mut(&c_inputs, cz_chunk);
-            });
-        drop(enter);
+        let span = tracing::span!(tracing::Level::DEBUG, "uniform_evals");
+        let _enter = span.enter();
+        let uni_constraint_evals: Vec<(Vec<(F, usize)>, Vec<(F, usize)>, Vec<(F, usize)>)> =
+            self.uniform_builder
+                .constraints
+                .par_iter()
+                .enumerate()
+                .map(|(constraint_index, constraint)| {
+                    let mut dense_output_buffer = unsafe_allocate_zero_vec(self.uniform_repeat);
+
+                    let mut evaluate_lc_chunk = |lc: &LC<I>| {
+                        if !lc.terms().is_empty() {
+                            let inputs = batch_inputs(lc);
+                            lc.evaluate_batch_mut(&inputs, &mut dense_output_buffer);
+
+                            // Take only the non-zero elements and represent them as sparse tuples (eval, dense_index)
+                            let mut sparse = Vec::with_capacity(self.uniform_repeat); // overshoot
+                            dense_output_buffer.iter().enumerate().for_each(
+                                |(local_index, item)| {
+                                    if !item.is_zero() {
+                                        let global_index =
+                                            constraint_index * self.uniform_repeat + local_index;
+                                        sparse.push((*item, global_index));
+                                    }
+                                },
+                            );
+                            sparse
+                        } else {
+                            vec![]
+                        }
+                    };
+
+                    let a_chunk: Vec<(F, usize)> = evaluate_lc_chunk(&constraint.a);
+                    let b_chunk: Vec<(F, usize)> = evaluate_lc_chunk(&constraint.b);
+                    let c_chunk: Vec<(F, usize)> = evaluate_lc_chunk(&constraint.c);
+
+                    (a_chunk, b_chunk, c_chunk)
+                })
+                .collect();
+
+        let (mut az_sparse, mut bz_sparse, cz_sparse) = par_flatten_triple(
+            uni_constraint_evals,
+            unsafe_allocate_sparse_zero_vec,
+            self.uniform_repeat, // Capacity overhead for offset_eq constraints.
+        );
```

Comment:
Definitely agree this is unsanitary. 

Unfortunately the plain `flat_map` approach adds about 60% overhead. 1.9s -> 3.3s on a 40s trace. I believe this is because it doesn't know the shape of the underlying and the chunks are very non-uniform sizes. `par_flatten_triple` is optimized to handle this well.

Ideally we'd know which `sparse_index` to write each non-zero LC eval to in advance, but the nature of sparsity is that we don't know. 

To get parallelism at the `evaluate_lc_chunk` level we compute `self.uniform_num_rows` LC computations in parallel. If we did this sparsely it would require a `.par_iter().map(...).collect()` during `evaluate_lc_chunk`, adding a painful allocation / copy.

The one approach that I have not tried is to chunk the constraints into `NUM_THREADS`-sized chunks, iterate over each chunk of constraints in parallel, then write directly to a sparse `Vec<(F, usize)>` directly for each chunk. We would then have to collect a `Vec<Vec<(F, usize)>> -> Vec<Vec<(F, usize)>>`.

In practice `num_constraints ~= 20`, thus the chunk size would be ~= 2. The expense of the LC computation is (very) non-uniform across constraints. We would lose parallelism across the LC (in order to append to a mutable vec) then get worse work stealing across constraints. We cannot account for this in the standard way by increasing the number of chunks as the chunk size would approach 1. There is theoretically a way to account for this by assigning a "computation weight" to each constraint / LC. I have not attempted this and imagine the code gets substantially grosser, but I imagine it could save 1% e2e.



Attaching the code for the flat_map version below so we can try it again later easily. Note the dense_output_buffers are over allocated in this version but it's not the driver of the performance decline.
```rust
        let evaluate_lc_chunk = |lc: &LC<I>,
                                     dense_output_buffer: &mut [F],
                                     constraint_index: usize|
         -> Vec<(F, usize)> {
            if !lc.terms().is_empty() {
                let inputs = batch_inputs(lc);
                lc.evaluate_batch_mut(&inputs, dense_output_buffer);

                // Take only the non-zero elements and represent them as sparse tuples (eval, dense_index)
                let mut sparse = Vec::with_capacity(self.uniform_repeat); // overshoot
                for (local_index, item) in dense_output_buffer.iter().enumerate() {
                    if !item.is_zero() {
                        let global_index = constraint_index * self.uniform_repeat + local_index;
                        sparse.push((*item, global_index));
                    }
                }
                sparse
            } else {
                vec![]
            }
        };

        // uniform_constraints: Xz[0..uniform_constraint_rows]
        let span = tracing::span!(tracing::Level::DEBUG, "uniform_evals");
        let _enter = span.enter();
        let mut az_sparse: Vec<(F, usize)> = self
            .uniform_builder
            .constraints
            .par_iter()
            .enumerate()
            .flat_map(|(constraint_index, constraint)| {
                let mut dense_output_buffer = unsafe_allocate_zero_vec(self.uniform_repeat);

                evaluate_lc_chunk(&constraint.a, &mut dense_output_buffer, constraint_index)
            })
            .collect();
        let mut bz_sparse: Vec<(F, usize)> = self
            .uniform_builder
            .constraints
            .par_iter()
            .enumerate()
            .flat_map(|(constraint_index, constraint)| {
                let mut dense_output_buffer = unsafe_allocate_zero_vec(self.uniform_repeat);

                evaluate_lc_chunk(&constraint.b, &mut dense_output_buffer, constraint_index)
            })
            .collect();
        let cz_sparse: Vec<(F, usize)> = self
            .uniform_builder
            .constraints
            .par_iter()
            .enumerate()
            .flat_map(|(constraint_index, constraint)| {
                let mut dense_output_buffer = unsafe_allocate_zero_vec(self.uniform_repeat);

                evaluate_lc_chunk(&constraint.c, &mut dense_output_buffer, constraint_index)
            })
            .collect();
        drop(_enter);
```

Reply:
Another idea: We could potentially nest flat_maps all the way up, but I believe this would have to be allocating / copying under the hood in a less controlled manner.